### PR TITLE
Get rid of the `Continue` IR node, as well as labels on loops.

### PR DESCRIPTION
**Pull request by sjrd**
Where we previously used `continue` statements in the IR, as in:
```javascript
lab: while (cond) {
  ...
  if (c)
    continue lab;
  ...
}
```
We now use a labeled block inside the loop, as well as a return off that labeled block, instead:
```javascript
while (cond) {
  lab[void]: {
    ...
    if (c)
      return(lab) undefined;
    ...
  }
}
```
This allows two important simplifications in the IR:

* We do not have the `Continue` IR node anymore.
* `While` and `DoWhile` loops do not have labels anymore.

With those changes, labels are only used by `Labeled` blocks and their corresponding `Return` statements.

To prevent some regressions in the quality of the generated .js code, we recover JS `continue` statements out of the above structure in the `FunctionEmitter`. We also received some *improvements* for free, because the `tailPosLabels` can kick in for those labels.

Full diff of the `testsuite-test-fastopt.js`: https://gist.github.com/sjrd/3d3f427d4b94072d4811b295c4c0a762

### CODE REVIEW

**Comment by sjrd** [comment]
I was working on the formalization of the IR for my thesis, and was bothered by the `continue`s, which I never *truly* understood myself. And then I realized I could get rid of them, and simplify a few things on the way 😄

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala [hunk 1]
```diff
@@ -575,54 +620,69 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
...
               unnest(cond) { (newCond, env0) =>
                 implicit val env = env0
-                js.If(transformExprNoChar(newCond),
-                    transformStat(body, Set.empty)(
-                        env.withDefaultBreakTargets(bodyBreakTargets)),
-                    js.Break())
+                js.If(transformExprNoChar(newCond), newBody, js.Break())
               }
-            }, newLabel)
+            })
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala]: Don't you need to use `optLabel` here? Otherwise how will the while get a label if it needs one?

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala [hunk 2]
```diff
@@ -575,54 +620,69 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
...
           }
 
-        case DoWhile(body, cond, label) =>
+        case DoWhile(body, cond) =>
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala]: Why do we not perform the same label optimization for `DoWhile`?

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala (hunk 1 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala]: Ouch, yes indeed!

I guess all the cases that are generated by Scala source code end up not needing a label ... which is why it wasn't caught by the CI. But in general, we might need one.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala (hunk 2 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala]: Because it's more complicated in a `DoWhile` due to the fact that the condition happens after the block. That means that in the case `!isExpression(cond)`, we cannot really do anything. In the case `isExpression(cond)`, we could. However, no Scala source code produces `do..while`s with this pattern, so the optimization would be dead code. (we get it for `while` loops--`while(true)` loops in particular--mostly because that's how we compile tail-recursive methods.

tl;dr, not worth it ;) I'll add a comment.

**Comment by sjrd** [comment]
Updated.
